torch/storage.py

```diff
@@ -40,6 +40,7 @@ def size(self) -> int:
     def type(self, dtype: _Optional[str] = None, non_blocking: bool = False) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
     def cuda(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
     def hpu(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
+    def xpu(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
```

Comment:
Ho why do we need these new methods? That's unfortunate :/

Reply:
> I would still like to discuss why we're adding new xpu() methods on the Storage object? We have all the generic equivalent there to satisfy any code update we might need to do right?

I am following the design architecture pattern which I observe here - add `xpu()` alongside with `cuda()` and `hpu()`. Note that at the moment you don't seem to have _the generic equivalent_ of these methods exposed on the uAPI level. So, with the current design I need to add `xpu()`. Alternate approach might be to expose actual generic uAPI, for example:
```
def backend(self, backend, device_idx=None, non_blocking=False, **kwargs) -> T: ... # type: ignore[empty-body, misc, type-var] # noqa: E704
```
and deprecate `cuda()` and `hpu()`. Eventually, we can add trivial wrappers on top of `backend` for backward compatibility.

Is above what you think about or you have deeper architecture refactoring in mind?

Reply:
> Note that at the moment you don't seem to have the generic equivalent of these methods exposed on the uAPI level.

Ho I missed that, that is indeed problematic here.

> Is above what you think about or you have deeper architecture refactoring in mind?

I would like to understand the difference between the backend() method you have and the .to() method we have on Tensor. They look very much the same at a high level.
If they are actually similar, I would go along the idea of https://github.com/pytorch/pytorch/blob/31372fa84293f5bdd0fb06acceb098a5c8187a07/torch/storage.py#L148-L154
But instead passing *args and **kwargs along to the call to to().
This will give us a generic API that is consistent with the Tensor API and that we don't need custom logic to maintain.

I think the 3 options we have here are:
1. Do this refactor to expose a full .to() API in this PR, use that and remove the xpu-specific method
2. Do the refactor in a follow up PR. Still remove the xpu-specific methods from this PR since we know we won't need them
3. Do the refactor in a follow up PR. Keep the xpu-specific APIs here knowing that we will deprecate and remove them once we add the generic API.

I personally think that 2 is the best option. If you do need a shorter-term solution/you won't be able to work on the refactor anytime soon, I could get convinced to do 3.

Reply:
We can't remove the xpu-specific methods from this PR, specifically - this `_StorageBase.xpu` method we are discussing here because it's essential to make a `load(map_location=xpu)` to work. Here is how it errors out on this PR if I comment xpu change in torch/storage.py:
```
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1014, in load
    return _load(opened_zipfile,
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1435, in _load
    result = unpickler.load()
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1405, in persistent_load
    typed_storage = load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1379, in load_tensor
    wrap_storage=restore_location(storage, location),
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1302, in restore_location
    return default_restore_location(storage, map_location)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 375, in default_restore_location
    result = fn(storage, location)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 347, in _deserialize
    raise RuntimeError(f'Attempting to load the storages to the {backend_name.upper()} device '
RuntimeError: Attempting to load the storages to the XPU device but torch.storage._StorageBase.xpu() or torch.storage.TypedStorage.xpu() is not generated. Please use torch.utils.generate_methods_for_privateuse1_backend to generate storage.xpu() method first.
```

So, I think that unfortunately option 2 is not possible and we need to decide on option 1 or option 3. That said I agree with your point that `_backend()` method I invented resembles `to()` and if expose uAPI, then it should be named `to()` and considered aligned with existing tensor's `to()`. While `_backend` is internal api, it's name is not that critical, but if exposed it should be `to()`.

I personally don't like to invent an API which will be deprecated next day. So, unless there will be a huge rock under the water, I suggest to look into option 1. Let's see how it will go for me today.

Let me know if I overlooked something in above estimation or exaggerated the problem.

Reply:
@albanD : ok, so I've just pushed one more commit which implements `.to()` for storage. Please, take a look if it goes in the right direction.